Support sub sequences in CharSequenceReader

robtimus opened the pull request:
Depends on PR 90.

garydgregory commented:
Thank you for the PR.

Some requests:

The Travis build fails with Checkstyle errors, may you fix these?

Create a Jira ticket and link it to this PR.

Update changes.xml with an entry for this PR.

Do the same for you other PR if it represents something differently as opposed to an evolution of this PR the user does not need to know about. 

Thank you!
Gary

robtimus commented:
See https://issues.apache.org/jira/browse/IO-619. I've added you as developer just like the other issues. I've also fixed a small bug in changes.xml where you referenced the wrong ticket number (2x 617 instead of 617 and 618).

robtimus commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -36,22 +36,105 @@
+     * Instead, if the character sequence grows smaller than the start index, this
+     * instance will act as if all characters have been read.
+     * </p>
+     *
```
I remembered I forgot those. I was already planning on adding them, you just beat me to it.

garydgregory commented:
Please rebase on master and omit changes.xml, it makes it easier to merge when dealing with lots of PRs. I should not have asked you to edit changes.xml in the first place! ;-)

robtimus commented:
Rebase complete, including revert of changes to changes.xml. That does mean that the incorrect 617 (should be 618) on line 131 is back.

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,107 @@
+     * @throws IllegalArgumentException if the start index is negative
+     * @since 2.7
+     */
+    public CharSequenceReader(final CharSequence charSequence, int start) {
```
`int start` -> `final int start`

robtimus replied:
Fixed

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,107 @@
+     * @throws IllegalArgumentException if the start index is negative, or if the end index is smaller than the start index
+     * @since 2.7
+     */
+    public CharSequenceReader(final CharSequence charSequence, int start, int end) {
```
Missing `final`

robtimus replied:
Fixed

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,107 @@
+    }
+
+    /**
+     * @return The start index in the character sequence, taking into account its length.
```
Missing Javadoc: "Starts..."

robtimus replied:
I've added a different comment as the method doesn't start anything, making it clearer what the method returns.

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,107 @@
+    }
+
+    /**
+     * @return The end index in the character sequence, taking into account its length.
```
Missing Javadoc: "Ends..."

robtimus replied:
I've added a different comment as the method doesn't end anything, making it clearer what the method returns.

garydgregory commented:
Hi @robtimus The PR is messed up now, it looks like you did a merge instead of a rebase. Please fix.

robtimus commented:
> Hi @robtimus The PR is messed up now, it looks like you did a merge instead of a rebase. Please fix.

Done. I ended up doing a cherry-pick from master instead, but the results are the same.

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,111 @@
     private int idx;
     private int mark;
 
+    /*
```
Move this Javadoc to#end and add @since 2.7

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,111 @@
+     * respectively. If end was an int, it would be initialized to 0 as well.
+     * That would cause all de-serialized CharSequenceReaders to be empty.
+     */
+    private final int start;
```
Give #start its own Javadoc and add @since 2.7.

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,111 @@
+    }
+
+    /**
+     * Construct a new instance with a portion of the specified character sequence.
```
"Construct" -> "Constructs"

garydgregory commented on src/main/java/org/apache/commons/io/input/CharSequenceReader.java:
```diff
@@ -37,22 +37,111 @@
+    }
+
+    /**
+     * Construct a new instance with a portion of the specified character sequence.
```
"Construct" -> "Constructs"
